prereq_tools/base.py

```diff
@@ -487,26 +487,61 @@ def update(self, subdir, **kw):
 class WebRetriever(object):
     """Identify a location from where to download a source package"""
 
-    def __init__(self, url):
+    def __init__(self, url, md5):
         self.url = url
+        self.md5 = md5
         self.__dry_run = GetOption('check_only')
         if self.__dry_run:
             SetOption('no_exec', True)
         self.__dry_run = GetOption('no_exec')
 
+    def check_md5(self, filename):
+        """Return True if md5 matches"""
+
+        if not os.path.exists(filename):
+            return False
+
+        with open(filename, "r") as src:
+            hexdigest = hashlib.md5(src.read()).hexdigest()
+
+        if hexdigest != self.md5:
+            print "Removing exising file %s: md5 %s != %s" % (filename,
+                                                              self.md5,
+                                                              hexdigest)
+            os.remove(filename)
+            return False
+
+        print "File %s matches md5 %s" % (filename, self.md5)
+        return True
+
+    def download(self, basename):
+        """Download the file"""
+        # Retry download a few times if it fails
+        for i in range(0, 3):
+            command = ['curl -L -O %s' % self.url]
+
+            if not RUNNER.run_commands(command):
+                print "Try #%d to get %s failed" % (i + 1, self.url)
+                continue
```

Comment:
Maybe a sleep in here to try to avoid doing all three attempts during a short outage?

Reply:
yeah, I thought of that but was lazy.  I just added a simple sleep with exponential backoff

Reply:
I rearranged it to do backoff in both cases